crates/wasm-encoder/src/component.rs

```diff
@@ -0,0 +1,371 @@
+
+use crate::{encoders, GlobalType, MemoryType, RawSection, Section, TableType};
+
+mod adapters;
+mod aliases;
+mod exports;
+mod functions;
+mod imports;
+mod instances;
+mod modules;
+mod types;
+
+pub use adapters::*;
+pub use aliases::*;
+pub use exports::*;
+pub use functions::*;
+pub use imports::*;
+pub use instances::*;
+pub use modules::*;
+pub use types::*;
+
+const INDEX_REF_INSTANCE: u8 = 0x00;
+const INDEX_REF_MODULE: u8 = 0x01;
+const INDEX_REF_FUNCTION: u8 = 0x02;
+const INDEX_REF_TABLE: u8 = 0x03;
+const INDEX_REF_MEMORY: u8 = 0x04;
+const INDEX_REF_GLOBAL: u8 = 0x05;
+
+const TYPE_REF_INSTANCE: u8 = 0x00;
+const TYPE_REF_MODULE: u8 = 0x01;
+const TYPE_REF_FUNCTION: u8 = 0x02;
+const TYPE_REF_TABLE: u8 = 0x03;
+const TYPE_REF_MEMORY: u8 = 0x04;
+const TYPE_REF_GLOBAL: u8 = 0x05;
+const TYPE_REF_ADAPTER_FUNCTION: u8 = 0x06;
+
+const CANONICAL_OPTION_UTF8: u8 = 0x01;
+const CANONICAL_OPTION_UTF16: u8 = 0x02;
+const CANONICAL_OPTION_COMPACT_UTF16: u8 = 0x03;
+const CANONICAL_OPTION_WITH_REALLOC: u8 = 0x04;
+const CANONICAL_OPTION_WITH_FREE: u8 = 0x05;
+
+/// A WebAssembly component section.
+///
+/// This trait marks sections that can be written to a `Component`.
+///
+/// Various builders defined in this crate already implement this trait, but you
+/// can also implement it yourself for your own custom section builders, or use
+/// `RawSection` to use a bunch of raw bytes as a section.
+pub trait ComponentSection {
+    /// This section's id.
+    ///
+    /// See `SectionId` for known section ids.
+    fn id(&self) -> u8;
+
+    /// Write this section's data and data length prefix into the given sink.
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>;
+}
+
+impl ComponentSection for RawSection<'_> {
+    fn id(&self) -> u8 {
+        self.id
+    }
+
+    fn encode<S>(&self, sink: &mut S)
+    where
+        S: Extend<u8>,
+    {
+        <Self as Section>::encode(self, sink);
+    }
+}
+
+/// A Wasm adapter module that is being encoded.
+#[derive(Clone, Debug)]
+pub struct AdapterModule {
+    bytes: Vec<u8>,
+}
+
+impl AdapterModule {
+    /// Begin writing a new `AdapterModule`.
+    pub fn new() -> Self {
+        Self {
+            bytes: vec![
+                0x00, 0x61, 0x73, 0x6D, // magic (`\0asm`)
+                0x0a, 0x00, 0x01, 0x00, // version
+            ],
+        }
+    }
+
+    /// Finish writing this adapter module and extract ownership of the encoded bytes.
+    pub fn finish(self) -> Vec<u8> {
+        self.bytes
+    }
+
+    /// Write a section into this adapter module.
+    pub fn section(&mut self, section: &impl ComponentSection) -> &mut Self {
+        self.bytes.push(section.id());
+        section.encode(&mut self.bytes);
+        self
+    }
+}
+
+impl Default for AdapterModule {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A Wasm component that is being encoded.
+#[derive(Clone, Debug)]
+pub struct Component {
```

Comment:
Oh I thought that adapter modules and components were one-and-the-same with components possibly being a refinement where no core items were exported, only interface-types-looking-items?

Reply:
I suspect you are right and `AdapterModule` will be deleted and one will be able to encode `Component`s in the module section along with core modules.

As it stands right now, the IT binary format proposal extends the updated module linking proposal's format, resulting in two different file formats (one for adapter modules from the linking spec and one for components from the IT spec).

This was the clarification I needed to get resolved before putting this up for review.

Reply:
So I talked more to Luke about this.

The updated spec for module linking will define its own module format to define "adapter modules" that may be used to adapt. This is what `AdapterModule` will be. It will support a subset of what components support in their sections.

Thus components may have three different types of modules encoded in its modules section: a core wasm module, a module adapter, or another component.

I'll refactor this a little to align with the proposal better.

Reply:
Hm ok I'm confused now. I was under the impression at the binary layer we'd only have adapter modules and core wasm modules, but from a viewpoint-perspective we'd also have components which just-so-happened to only import/export interface-type looking things. 

I can open a separate issue for this, though. I didn't know though that there was a proposal yet for the binary encoding of interface types?

Reply:
It's a work in progress in this branch: https://github.com/WebAssembly/interface-types/blob/component-rebase/design/proposals/interface-types/Binary.md

Reply:
I've pushed a new `adapter` module that defines `AdapterModule` and its relevant sections based on the binary format outlined in the proposal link above.

It doesn't currently share much code with the `component` module as the differences between the two formats can be subtle so sharing sections could be problematic.